src/nm-keyfile.c

```diff
@@ -138,32 +152,76 @@ netplan_render_yaml_from_nm_keyfile(GKeyFile* kf, const char* rootdir)
     }
     nd_type = type_from_str(type);
 
+    /* Use previously existing netdef IDs, if available, to override connections
+     * Else: generate a "NM-<UUID>" ID */
+    if (netdef_id)
+        nd_id = g_strdup(netdef_id);
+    else
+        nd_id = g_strconcat("NM-", uuid, NULL);
```

Comment:
The problem with this is, that NM does not have an understanding of a "Netplan Netdef ID". In Netplan this can be an pretty arbitrary string/identifier, but usually it is the name of the interface. Furthermore, NM does not set a specific `interface-name` by default, but rather creates connections per connection type (ethernet, wifi, vxlan, ...), which it will then apply to any matching interface.

So when a new connection is created via NM (i.e. `nmcli con add type ethernet`) we're in a situation where we neither have a interface name, nor any pre-existing Netplan Netdef ID. In this case we fallback to a `NM-<UUID>` netdef id.

For consistency, the YAML filename will always be `90-NM-<UUID>.yaml`, as this way its origin can easily be detected by NM and/or netplan and it is always the same, as every NM connection profile will have a UUID.

The NM keyfile filename on the other hand, is generated by `nm.c` and follows the structure: `netplan-<NETDEF_ID>[-<SSID>].nmconnection`. So this can be something like `netplan-eth0.nmconnection` or `netplan-NM-SOME_UUID-SOME_SSID.nmconnection`. When a existing connection is modified via NM (i.e. `nmcli con mod netplan-eth0 connection.id myconnection`), it will extract the previous netdef id from the keyfile filename and pass it to this function, so that the newly created netdef can use the same ID, and overwrite the previous connection from the YAML, to avoid rendering the same connection twice under different names/IDs.

The naming of Netdef/YAML/Keyfile is quite complex but cannot be unified, due to missing information  (like the netdef ID in NM). But I hope this clarifies the situation a bit!